edgraph/server.go

```diff
@@ -1894,6 +1894,218 @@ func (s *Server) CheckVersion(ctx context.Context, c *api.Check) (v *api.Version
 	return v, nil
 }
 
+func (s *Server) CreateNamespace(ctx context.Context, in *api.CreateNamespaceRequest) (
+	*api.CreateNamespaceResponse, error) {
+
+	if err := AuthGuardianOfTheGalaxy(ctx); err != nil {
+		s := status.Convert(err)
+		return nil, status.Error(s.Code(),
+			"Non guardian of galaxy user cannot create namespace. "+s.Message())
+	}
+
+	if err := verifyNamespaceName(in.NsName); err != nil {
+		return nil, err
+	}
+
+	if _, err := getNamespaceID(ctx, in.NsName); err == nil {
+		return nil, errors.Errorf("namespace %q already exists", in.NsName)
+	} else if !strings.Contains(err.Error(), "not found") {
+		return nil, err
+	}
+
+	password := "password"
+	if len(in.Password) != 0 {
+		password = in.Password
+	}
+
+	ns, err := (&Server{}).CreateNamespaceInternal(ctx, password)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we crash at this point, it is possible that namespaces is created
+	// but no entry has been added to dgraph.namespace predicate. This is alright
+	// because we have not let the user know that namespace has been created.
+	// The user would have to try again and another namespace then would be
+	// assigned to the provided name here.
+
+	_, err = (&Server{}).QueryNoGrpc(ctx, &api.Request{
+		Mutations: []*api.Mutation{{
+			Set: []*api.NQuad{
+				{
+					Subject:     "_:ns",
+					Predicate:   "dgraph.namespace.name",
+					ObjectValue: &api.Value{Val: &api.Value_StrVal{StrVal: in.NsName}},
+				},
+				{
+					Subject:     "_:ns",
+					Predicate:   "dgraph.namespace.id",
+					ObjectValue: &api.Value{Val: &api.Value_IntVal{IntVal: int64(ns)}},
+				},
+			},
+		}},
+		CommitNow: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	glog.Infof("Created namespace [%v] with id [%d]", in.NsName, ns)
+	return &api.CreateNamespaceResponse{NsId: ns}, nil
+}
+
+func (s *Server) DropNamespace(ctx context.Context, in *api.DropNamespaceRequest) (
+	*api.DropNamespaceResponse, error) {
```

Comment:
Do we need response if there's nothing in it?

Reply:
This is usually a good idea for future extensibility so that we can modify the response if needed. This will be mostly hidden from the user through the client APIs that we expose.